### PR TITLE
Add option to automatically handle unsorted variable-length sequences in RNNs

### LINKED ISSUE #3584: Sort sequences internally in pack_padded_sequence
Currently, `pack_padded_sequence` requires the caller to sort `inputs` by the length (descending). We should sort and reorder the input internally if it's not already sorted. We can then restore the original order in `pad_packed_sequence`.

cc @kyunghyuncho

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #3584.

Motivation: manually sorting sequences, packing them, and then unsorting them
is something a lot of users have complained about doing, especially when we can
offer library support for them.

Overview: we internally sort sequences before packing them and store a list of
`unsorted_indices` that represent how to unsort the sequences inside
PackedSequence. The packing helper functions return PackedSequence with the
`permutation` field and the unpacking helper functions use it to unsort.

To implement this, the following changes were made:
- PackedSequence now keeps `sorted_indices` and `unsorted_indices`.
  These two can be thought of as permutations and are inverses of each other.
  `sorted_indices` is how the sequences were sorted; `unsorted_indices` is how
  to unsort the sequences.
- Added an `enforce_sorted` argument to pack_sequence and pack_padded_sequence
  that maintains the legacy behavior of error-ing out on unsorted-sequences.
  When `enforce_sorted=True`, these functions maintain their ONNX exportability.
- pack_sequence(sequences, enforce_sorted) takes in unsorted sequences.
- pack_padded_sequence can take in a padded tensor that represents padded,
  unsorted sequences.
- pad_packed_sequence unsorts the PackedSequence such that it is still the
  inverse operation of packed_padded_sequence.
- RNNs apply `sort_indices` to their input hidden state and apply
  `unsort_indices` to their output hidden state. This is to ensure that the
  hidden state batches correspond to the user's ordering of input sequences.

NOT BC-Breaking
- The default for pack_sequence and pack_padded_sequence is
  `enforce_sorted=True` to avoid breaking ONNX export. To use the new
  functionality, pass in `enforce_sorted=False`

Testing Plan
- Modified TestNN.test_pack_sequence, TestNN.test_packed_padded_sequence,
  and TestNN.test_variable_sequence (RNN test) to check the behavior
  of unsorted sequences, sorted sequences, and sorted sequences with
  enforce_sorted=True
- test/test_jit.py has a test to see if RNNs are exportable with
  enforce_sorted=True

cc @colesbury

### CODE REVIEW

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -18,6 +18,13 @@
 }
 
 
+def apply_permutation(tensor, permutation, contiguous_output, dim=1):
+    result = tensor.index_select(dim, permutation)
+    if contiguous_output:
```
Comment (fmassa) [inline torch/nn/modules/rnn.py]: Question: don't we always return contiguous tensors from index_select?

Comment (zou3519) [reply torch/nn/modules/rnn.py]: oh, you have a good point, I forgot about that

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -171,6 +185,10 @@ def forward(self, input, hx=None):
                                  requires_grad=False)
             if self.mode == 'LSTM':
                 hx = (hx, hx)
+        else:
+            # Each batch of the hidden state should match the input sequence that
+            # the user believes he/she is passing in.
+            self.permute_hidden(hx, sorted_indices)
```
Comment (colesbury) [inline torch/nn/modules/rnn.py]: I'm confused by this line. It looks like `permute_hidden` returns the result and does not modify its inputs, bu this line doesn't use the output.

Comment (zou3519) [reply torch/nn/modules/rnn.py]: Good catch -- it should be `hx = self.permute_hidden`, I will update that

Comment (zou3519) [reply torch/nn/modules/rnn.py]: Fixed and added a test for this case.

**Review by colesbury — APPROVED** [review]
This lgtm, but I'm not super familiar with the rnn code so it might be useful to get someone that knows it better to look over it too.

**Pull request closed without merging**